Implement new swift_proto_library rule (#1140)
bazelbuild/rules_swift: doc/proto_migration.md

# Overview

This document aims to provide context to future contributors on why we decided to rewrite

> the swift_proto_library and swift_grpc_library rules into the new swift_proto_library rule,

Author: Sure, I went through and updated the migration doc to improve the formatting.

## 2. swift_grpc_library to new_swift_proto_library

Given the following proto_library target where echo.proto contains the definition
for the Echo service with an RPC named Echo which takes an EchoRequest and returns an EchoResponse:

```
proto_library(
    name = "echo_proto",
    srcs = ["echo.proto"],
)
```

With the following deprecated swift_proto_library and swift_grpc_library targets:

```
swift_proto_library(
    name = "echo_proto_swift",
    deps = [":echo_proto"],
)

swift_grpc_library(
    name = "echo_client_services_swift",
    srcs = [":echo_proto"],
    flavor = "client",
    deps = [":echo_proto_swift"],
)

swift_grpc_library(
    name = "echo_client_test_stubs_swift",
    srcs = [":echo_proto"],
    flavor = "client_stubs",
    deps = [":echo_client_services_swift"],
)

swift_grpc_library(
    name = "echo_server_services_swift",
    srcs = [":echo_proto"],
    flavor = "server",
    deps = [":echo_proto_swift"],
)
```

These can be migrated to the following new_swift_proto_library targets:

```
new_swift_proto_library(
    name = "service_server_swift_proto",

>     compilers = ["//swift/proto/compilers:swift_server_proto"],

Reviewer: Should this include the repository name ?

```
    compilers = ["@rules_swift//swift/proto/compilers:swift_server_proto"],
```

Author: @cgrindel If they're using the rules_swift proto compilers, yeah, I can add that. Though in the rules_swift_package_manager GRPC example, I demonstrated creating a custom compiler to use the swiftpkg GRPC / SwiftProtobuf dependencies which didn't need a workspace name because it was in the same workspace.

# Overview

This document aims to provide context to future contributors on why we decided to rewrite
the `swift_proto_library` and `swift_grpc_library` rules into the new `swift_proto_library` rule,
as well as a guide for consumers of the deprecated rules to migrate to the new `swift_proto_library` rule.

On this page:
  * [Why rewrite?](#why-rewrite)
  * [How to Migrate](#how-to-migrate)

## Why rewrite?

The new `swift_proto_library` rule was created to address several issues with the old
`swift_proto_library` and `swift_grpc_library` rules which would be difficult if not impossible 
to fix while maintaining full backwards compatibility.

### 1. The Swift Protoc Gen Aspect

Firstly, the old `swift_proto_library` rule used the `swift_protoc_gen_aspect` aspect
which traversed its direct `proto_library` dependencies and their transitive graph of `proto_library` dependencies,
and compiled all of these into swift modules. 

This meant that a single `swift_proto_library` target could generate and and compile swift source files
for *all* of the `proto_library` targets in potentially a very large dependency graph,
which could span across multiple repositories.

> In practice, though, consumers typically preferred to have a 1:1 mapping of `proto_library` to `swift_proto_library` targets.
> Additionally, this aspect meant that the module name and other compilation options could not be configured on a per-target basis,
> since the aspect needed to be able to determine all of this information by just looking at the providers of the `proto_library` targets.

Reviewer 2: ```suggestion
In practice, though, consumers typically preferred to have more control over the mapping of `proto_library` to `swift_proto_library` targets.
Additionally, this aspect meant that the module name and other compilation options could not be configured on a per-target basis,
since the aspect needed to be able to determine all of this information by just looking at the providers of the `proto_library` targets.
```

> This lack of flexibility was extremely frustrating, especially for consumers who wanted to assign "Swift-like" module names
> to the `swift_proto_library` modules. 

Reviewer 2: nit: imo think the point is made in the previous and next paragraph.

```suggestion
```

they can assign an arbitrary module name to the target and import it with:

```
import ServiceServer
```

This removes the guesswork of trying to determine what the module name will be,
and it allows engineers to work with the new `swift_proto_library` targets exactly the same
way they would if these were `swift_library` targets they had written by hand,
except the source files are generated for them.

### 2. Implicit, internal proto compiler and runtime dependencies.

The second main issue the new implementation is intended to address is the specific proto compiler,
proto compiler plugin, and compile time proto library dependencies the old rules required.

The old rules depended internally on the SwiftProtobuf and GRPC targets defined in the targets under
the `third_party` directory. This was problematic for consumers who wanted to use different versions
of protoc, the protoc plugins or compile time dependencies.

Specifically, this issue came to a head when implementing support for proto targets in
`rules_swift_package_manager`. There was a lengthy slack conversation here on the topic:
https://bazelbuild.slack.com/archives/CD3QY5C2X/p1692055426375909

> The problem was that some Swift packages like Firebase have package dependencies on protobuf
> libraries which contain the same code (though perhaps a different version) as the protobuf
> targets defined in this repository which were dependencies of those rules.

Reviewer 2: ```suggestion
The problem was that some Swift packages have package dependencies on protobuf
libraries which contain the same code (though perhaps a different version) as the protobuf
targets defined in this repository which were dependencies of those rules.
```

        "ExtraModuleImports": "examples_xplatform_grpc_service_client_swift_proto",
    },
    compiler_deps = [
        ":service_client_swift_proto",
    ],
    compilers = ["@build_bazel_rules_swift//proto/compilers:swift_test_client_proto"],
    protos = [":service_proto"],
)
```

Note here that we don't need the intermediate `swift_proto_library` target, 
and that we must pass a `swift_proto_compiler` target via the compilers attribute
which is configured with a `grpc-swift` protoc plugin.

Also note that the `client_stubs` flavor uses the `swift_test_client_proto` grpc compiler
and requires additional plugin options and compiler dependencies.

This configuration was added to ensure continuity of rule capabilities,
but when possible we recommend migrating off of these targets. 
This is because TestClient protoc plugin option was deprecated with Swift 5.6,
as the test clients do not conform to Sendable.
This capability may be removed in a future major version update.

The alternative recommended by the `grpc-swift` authors is to register a mock server on localhost.

## F.A.Q.

As consumers raise questions during their migrations to the new `swift_proto_library` rule,

> we will add their answers here as a reference.

Reviewer 2: Some I can think of (and potentially we just link to examples here):

- Can I group protos to reduce number of modules?
- Can I use a completely custom protobuf compiler?

Author: Added answers to these questions
